### PR TITLE
introducing ICNT86Touchscreen

### PR DESCRIPTION
**Pull request by siemon-geeroms**
# What does this implement/fix?

Adding the touch driver for[Pico-CapTouch-ePaper-2.9](https://www.waveshare.com/wiki/Pico-CapTouch-ePaper-2.9)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
touchscreen:
  - platform: icnt86 
    interrupt_pin: GPIO17
    reset_pin: GPIO16
    on_touch: 
      then:
        - component.update: my_display
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nielsnl68** [comment]
Just added a ping moment for when this is merged before my touchscreen PR https://github.com/esphome/esphome/pull/4596 is merged.

File: esphome/components/icnt86/icnt86.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+float ICNT86Touchscreen::get_setup_priority() const { return setup_priority::HARDWARE - 1.0f; }
+
+void ICNT86Touchscreen::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up icnt86 Touchscreen...");
+
+  // Register interrupt pin
+  this->interrupt_pin_->pin_mode(gpio::FLAG_INPUT | gpio::FLAG_PULLUP);
+  this->interrupt_pin_->setup();
+  this->store_.pin = this->interrupt_pin_->to_isr();
```
Comment (nielsnl68) [inline esphome/components/icnt86/icnt86.cpp:21]:
```suggestion
```

File: esphome/components/icnt86/icnt86.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    this->reset_();
+  }
+
+  // Update display dimensions if they were updated during display setup
+  this->display_width_ = this->display_->get_width();
+  this->display_height_ = this->display_->get_height();
+  this->rotation_ = static_cast<TouchRotation>(this->display_->get_rotation());
+
+  // Trigger initial read to activate the interrupt
+  this->store_.touch = true;
```
Comment (nielsnl68) [inline esphome/components/icnt86/icnt86.cpp:36]:
```suggestion
```

File: esphome/components/icnt86/icnt86.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#define UBYTE uint8_t
+#define UWORD uint16_t
+#define UDOUBLE uint32_t
+
+struct ICNT86TouchscreenStore {
+  volatile bool touch;
+  ISRInternalGPIOPin pin;
+  uint8_t touch_count;
+  static void gpio_intr(ICNT86TouchscreenStore *store);
+};
```
Comment (nielsnl68) [inline esphome/components/icnt86/icnt86.h:22]:
```suggestion
```

File: esphome/components/icnt86/icnt86.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ protected:
+  void reset_();
+  void I2C_Write_Byte_(UWORD Reg, char *Data, UBYTE len);
+  void ICNT_Read_(UWORD Reg, char *Data, UBYTE len);
+  void ICNT_Write_(UWORD Reg, char *Data, UBYTE len);
+  void I2C_Read_Byte_(UWORD Reg, char *Data, UBYTE len);
+  void ICNT_ReadVersion_();
+  void reset_touch_sensor_();
+
+  ICNT86TouchscreenStore store_;
```
Comment (nielsnl68) [inline esphome/components/icnt86/icnt86.h:43]:
```suggestion
```

**Review by nielsnl68 — COMMENTED** [review]
You was almost there. Why did you close the PR?

**Comment by siemon-geeroms** [comment]
> You was almost there. Why did you close the PR?

This was by accident created a new https://github.com/esphome/esphome/pull/5960

**Pull request closed without merging**